src/frontend/parser.mly

```diff
@@ -375,6 +389,18 @@ remaining_declarations(rhs):
  * identifier.
  *)
 decl(type_rule, rhs):
+  (* This is just a helper for the fact that we don't support the old array syntax any more *)
+  | ty=type_rule id=decl_identifier LBRACK {
+    raise
+    (Errors.SyntaxError
+       (Errors.Parsing
+          ( Fmt.str
+              "\";\" expected after variable declaration.@ It looks like you \
+               are trying to use the old array syntax.@ Please use the new \
+               syntax:@ @[<h>array[...] %a %s;@]@\n"
```

Comment:
I don't think this needs the generic "expected semicolon" part since the specific error is identified quite precisely.
Also, it wouldn't be too difficult to print the exact array dimensions instead of `[...]`:
```suggestion
  | ty=type_rule id=decl_identifier LBRACK dims=separated_nonempty_list(COMMA, expression) RBRACK {
    let (ty, trans) = ty in
    let ty = List.fold_right ~f:(fun e ty -> SizedType.SArray (ty, e)) ~init:ty dims in
    let ty = (ty, trans) in
    raise
    (Errors.SyntaxError
       (Errors.Parsing
          ( Fmt.str
              "Deprecated array syntax.@ Please use the new \
               syntax:@ @[<h>%a %s;@]@\n"
```
And actually, since this doesn't cause parsing ambiguities, maybe the canonicalizer could just accept the old array syntax.

Reply:
> I don't think this needs the generic "expected semicolon" part since the specific error is identified quite precisely

I left it in as a guard for there being a genuine mistake where the user didn't want an array at all but had a stray character which happened to be a `[`. It seemed odd that any other special character there would mention a `;`, but this wouldn't.

> maybe the canonicalizer could just accept the old array syntax.

It would need to be a bit more complicated; if you wanted to really allow that you'd also need to support the optional assignment (essentially go back to the parsing rule pre-#1388). I don't think this would cause any conflicts at the moment either.

I don't want to do that, mainly because as long as the canonicalizer would support it, some interfaces would just update their pipeline to `canonicalize -> stanc -> c++ compile` and therefore never actually update the Stan code. That would "work" unless/until we actually wanted to do something syntactically which was incompatible, at which point we're either back to where we are now, or our hands remain tied and the deprecation policy was not able to achieve the main goal of allowing language evolution.

Reply:
> a genuine mistake where the user didn't want an array at all but had a stray character

That seems unlikely, and I think that as long as the caret points to the errant bracket, the user can spot the mistake easily even if the error message is nonsense.
Also, this problem disappears if you go with my suggestion of parsing the whole dimension expression.

> some interfaces would just update their pipeline to canonicalize -> stanc -> c++ compile and therefore never actually update the Stan code. 

Is this a hypothetical, or do you know some interface that opposes our deprecation policy?

Reply:
I know BRMS was using that set up until 2.34 forced them not to. I do not know of any others concretely, but I would be entirely unsurprised if some other downstream users of RStan have done this

I made the change to give a complete suggestion. It introduced a few spurious error cases in the messages file for situations where you have `int foo[` followed by an invalid expression - I gave them all a generic message in that case.